rust/noosphere-gateway/src/gateway_manager.rs

```diff
@@ -0,0 +1,39 @@
+use anyhow::Result;
+use async_trait::async_trait;
+use axum::http::{request::Parts, StatusCode};
+use noosphere_core::{context::HasMutableSphereContext, data::Did};
+use noosphere_storage::Storage;
+use std::pin::Pin;
+use tokio_stream::Stream;
+
+#[cfg(doc)]
+use noosphere_core::context::SphereContext;
+
+/// [Stream] of [SphereContext] from a [GatewayManager].
+pub type GatewayManagerSphereStream<'a, C> = dyn Stream<Item = Result<C>> + Send + 'a;
+
+/// [GatewayManager] implementations are used to provide access to managed
+/// hosted sphere data and customizations in a Noosphere gateway.
+#[async_trait]
+pub trait GatewayManager<C, S>: Clone + Send + Sync
+where
+    C: HasMutableSphereContext<S>,
+    S: Storage + 'static,
+{
+    /// Retrieve a sphere context whose counterpart matches `counterpart`.
+    async fn get_sphere_context(&self, counterpart: &Did) -> Result<C>;
+
+    /// Retrieve the identity of the managed sphere's device key from provided `counterpart`,
+    /// as a lightweight alternative to fetching the entire sphere context
+    /// in [GatewayManager::get_sphere_context].
+    async fn get_gateway_identity(&self, counterpart: &Did) -> Result<Did>;
+
+    /// Iterate over all managed spheres. Currently used by task workers.
+    /// Will need to rethink this.
```

Comment:
Aye the GatewayManager should never be accessed under a route handler, outside of the middleware parsing which pulls out the relevant scope information. The `iter()` function is only used as a stop-gap for current embedded workers, and is conflated with `GatewayWorker` currently for migration ease. This should go away once migrating to a new vision of workers -- will clarify those plans in a follow up issue.

In the mean time, there may be some Axumisms we can use to prevent pulling this in as a router arg, or some scary renaming e.g. `.experimental_worker_only_iter()`